docs/model-dev-guide/dtrain/dtrain-implement.rst

```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
+feature, ``slots_per_trial`` should be set as a multiple of the total number of GPUs on an agent
+machine. For example, if your resource pool consists of multiple 8-GPU agent machines, valid
+``slots_per_trial`` values would be 16, 24, 32, and so on.
+
+In the following configuration, trials will use the combined resources of multiple machines to train
+a model:
 
 .. code:: yaml
 
    resources:
      slots_per_trial: 16  # Two 8-GPU agent machines will be used in a trial
 
-For distributed multi-machine training, Determined automatically detects a common network interface
-shared by the agent machines. If your cluster has multiple common network interfaces, please specify
-the fastest one in :ref:`cluster-configuration` under
+For distributed multi-machine training, Determined will automatically detect a common network
+interface that is shared by the agent machines. If your cluster has multiple common network
+interfaces, we advise specifying the fastest one in :ref:`cluster-configuration` under
 ``task_container_defaults.dtrain_network_interface``.
 
-Reduce computational overhead by setting the ``global_batch_size`` to the largest batch size that
-fits into a single GPU multiplied times the number of slots.
+You can reduce computational overhead by setting the ``global_batch_size`` to the largest batch size
+that fits into a single GPU multiplied times the number of slots.
 
 When the ``slots_per_trial`` field is set, the per-slot (i.e., per-GPU) batch size is set to
-``global_batch_size // slots_per_trial``. The per-slot and global batch sizes should be accessed via
-the context using :func:`context.get_per_slot_batch_size()
+``global_batch_size // slots_per_trial``. The per-slot and global batch sizes can be accessed
+through the context using :func:`context.get_per_slot_batch_size()
 <determined.TrialContext.get_per_slot_batch_size>` and :func:`context.get_global_batch_size()
 <determined.TrialContext.get_global_batch_size>`, respectively. If ``global_batch_size`` is not
 evenly divisible by ``slots_per_trial``, the remainder is dropped.
 
-If :ref:`slots_per_trial <exp-config-resources-slots-per-trial>` is greater than the number of slots
-on a single agent, Determined schedules it over multiple machines. When scheduling a multi-machine
-distributed training job, Determined requires that the job uses all of the slots (GPUs) on an agent.
-For example, in a cluster that consists of 8-GPU agents, an experiment with :ref:`slots_per_trial
-<exp-config-resources-slots-per-trial>` set to ``12`` is never scheduled and will wait indefinitely.
-The section on :ref:`Scheduling Behavior <dtrain-scheduling>` describes this in more detail.
+If the value of :ref:`slots_per_trial <exp-config-resources-slots-per-trial>` is greater than the
+number of slots available on a single agent, Determined schedules it over multiple machines. When
+scheduling a multi-machine distributed training job, Determined requires that all slots (GPUs) on an
```

Comment:
should this info about uneven slots be a separate note? not sure what our standard is for note/warning sections. but seems like something worth calling out, slightly unexpected behavior.

Reply:
@azhou-determined is the uneven slot constraint still true? I thought we changed the scheduler behavior so that you could use uneven slots. Or at least had some experimental feature that could do so.

I also think this paragraph is a little redundant with the earlier paragraph about multi-machine parallelism. Or at least the two paragraphs could be merged.

Reply:
answer:

re: "When scheduling a multi-machine distributed training job, Determined requires that the job uses all of the slots (GPUs) on an agent. For example, in a cluster that consists of 8-GPU agents, an experiment with [slots_per_trial](https://docs.determined.ai/latest/reference/training/experiment-config-reference.html#exp-config-resources-slots-per-trial) set to 12 is never scheduled and will wait indefinitely."

Determined now supports uneven slots, see the Scheduler setting "allow_heterogeneous_fits" documented in the master config doc: https://docs.determined.ai/latest/reference/deploy/config/master-config-reference.html#allow-heterogeneous-fits

Reply:
resolved